lib/grandpa/types.go

```diff
@@ -40,6 +43,42 @@ type State struct {
 	round  uint64   // voting round number
 }
 
+// NewState returns a new GRANDPA state
+func NewState(voters []*Voter, setID, round uint64) *State {
+	return &State{
+		voters: voters,
+		setID:  setID,
+		round:  round,
+	}
+}
+
+// pubkeyToVoter returns a Voter given a public key
+func (s *State) pubkeyToVoter(pk *ed25519.PublicKey) (*Voter, error) {
+	max := uint64(2^64) - 1
+	id := max
+
+	for i, v := range s.voters {
+		if bytes.Equal(pk.Encode(), v.key.Encode()) {
+			id = uint64(i)
+			break
+		}
+	}
+
+	if id == max {
+		return nil, ErrVoterNotFound
+	}
+
+	return &Voter{
+		key: pk,
+		id:  id,
+	}, nil
+}
+
+// threshold returns the 2/3 |voters| threshold value
+func (s *State) threshold() uint64 {
+	return uint64(2 * len(s.voters) / 3)
+}
+
```

Comment:
Should this be truncated?  or rounded?  In the case where s.voters = 6, the result is 4.  And when there are 7 voters the result is 4, but 4 is not greater than 2/3 * 7 which is 4.666, should this instead be rounded to 5 which is greater than 2/3 * 7?  Is this addressed in the spec?

Reply:
yeah this is a good question, I left it rounded down for now since in practice the validator set should be ~100 so it wouldn't make a huge difference. I can ask and clarify though

Reply:
Would be nice to add a comment/note here in the code or a todo to make sure we loop back.

Reply:
I've added a todo for now